front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/v2/ApplicationsControllerV2.groovy

```diff
@@ -0,0 +1,163 @@
+  void delete(@PathVariable String application, HttpServletResponse response) {
+    getApplication().initialize(new Application().withName(application)).delete()
+    response.setStatus(HttpStatus.ACCEPTED.value())
+  }
+
+  @PreAuthorize("hasPermission(#app.name, 'APPLICATION', 'WRITE')")
+  @ApiOperation(value = "", notes = "Update an existing application")
+  @RequestMapping(method = RequestMethod.PUT, value = "/{name}")
+  Application update(@PathVariable String name, @RequestBody final Application app) {
+    if (name.trim() != app.getName()) {
+      log.error("non matching application name {}", name)
+      throw new InvalidApplicationRequestException("provided application name does not match")
+    }
+
+    def application = getApplication()
+    Application existingApplication = application.findByName(app.getName())
+    application.initialize(existingApplication).withName(app.getName()).update(app)
+    return app
+  }
+
+  @PreAuthorize("hasPermission(#application, 'APPLICATION', 'READ')")
+  @ApiOperation(value = "", notes = "Fetch a single application by name")
+  @RequestMapping(method = RequestMethod.GET, value = "/{application}")
+  Application get(@PathVariable final String application) {
+    if (!application?.trim()) {
+      throw new NotFoundException("No application name provided")
+    }
+
+    return applicationDAO.findByName(application.toUpperCase())
+  }
+
+  @PreAuthorize("hasPermission(#application, 'APPLICATION', 'READ')")
+  @RequestMapping(value = '{application}/history', method = RequestMethod.GET)
+  Collection<Application> getHistory(@PathVariable String application,
+                                     @RequestParam(value = "limit", defaultValue = "20") int limit) {
+    return applicationDAO.getApplicationHistory(application, limit)
+  }
+
+  @PreAuthorize("@fiatPermissionEvaluator.isAdmin()")
+  @RequestMapping(method = RequestMethod.POST, value = "/batchUpdate")
+  void batchUpdate(@RequestBody final Collection<Application> applications) {
+    applicationDAO.bulkImport(applications)
+  }
+
+  @ExceptionHandler(Application.ValidationException)
+  @ResponseStatus(HttpStatus.BAD_REQUEST)
+  Map handleValidationException(Application.ValidationException ex) {
+    def locale = LocaleContextHolder.locale
+    def errorStrings = []
+    ex.errors.each { Errors errors ->
+      errors.allErrors.each { ObjectError objectError ->
+        def message = messageSource.getMessage(objectError.code, objectError.arguments, objectError.defaultMessage, locale)
+        errorStrings << message
+      }
+    }
+    return [error: "Validation Failed.", errors: errorStrings, status: HttpStatus.BAD_REQUEST]
+  }
+
+  @ExceptionHandler(AccessDeniedException)
+  @ResponseStatus(HttpStatus.FORBIDDEN)
+  Map handleAccessDeniedException(AccessDeniedException ade) {
+    return [error: "Access is denied", status: HttpStatus.FORBIDDEN.value()]
+  }
+
+  private Application getApplication() {
+    return new Application(
```

Comment:
Yeah I wanted to keep the change small for this commit. The change was going to be just a bit big but we can yank it out.